resources/tests/org.eclipse.core.tests.resources/src/org/eclipse/core/tests/resources/content/LazyReaderTest.java

```diff
@@ -137,4 +150,35 @@ public void testMarkAndReset() throws IOException {
 		assertEquals("3.0", 0, stream.getOffset());
 		stream.close();
 	}
+
+	@Test
+	public void testEnsureAvailable_BufferSizeDoesNotOverflow() throws IOException {
+		Reader infinitelyEmpty = new Reader() {
+
+			@Override
+			public int read() throws IOException {
+				return 0;
+			}
+
+			@Override
+			public int read(char[] b, int off, int len) throws IOException {
+				return len;
+			}
+
+			@Override
+			public void close() throws IOException {
+				// nothing to close
+			}
+		};
+
+		try (OpenLazyReader objectUnderTest = new OpenLazyReader(infinitelyEmpty, 10)) {
+			// HACK: needed to avoid filling up the RAM with 0's when calling "skip"
+			objectUnderTest.setBufferSize(Integer.MAX_VALUE);
+			objectUnderTest.setOffset(Integer.MAX_VALUE);
```

Comment:
Please do not add further "for testing purposes only" methods for the class. If a class cannot reach a state of interest, such as a faulty one for a test, by using its API, there is not need to test that case. And in case the state can be achieved via its API, there is no need to add further methods to make testing more convenient.

In this case the used `Reader` mock can, for example, track the number of `read()` calls to identify the situation of interest.

Reply:
Not possible: `read` would have to be called so often to get to the interesting edge case that the heap would overflow. The trick is to either hack the `LazyReader` so it doesn't load 2^32 elements into an array. In order to do that all I could come up with was either:
a. Hacking the `bufferSize` and the `offset` (what I did) or 
b. I could declare `blockCapacity` and `loadBlock()` as `protected` and avoid filling up the array in the `OpenLazyReader`. 

Do you think option "b" would be better?

If not, how would you code it?

Reply:
So this only because the tests run with 2 GB heap space. There was a recent PR increasing that to 4 GB, so maybe it will work now?
eclipse-platform/eclipse.platform.releng.aggregator#1612